### ISSUE TITLE
Canonical principal name was established as 'null'

### ISSUE THREAD

**Issue by gegmar** [Ratler/shibboleth-mfa-u2f-auth#2]
Hi again!

I have set up my shibboleth as you described in your README.
- Java1.8.0_v74
- Tomcat 8.0.32
- Shibboleth 3.2.1
- HEAD/Master of this Project

Everything works as expected: I am able to login with username/password and then with my U2F-Device.
But after the authentication, the verified username will not be given to the core-application.
With `<variable name="idp.loglevel.idp" value="DEBUG" />` in `conf/logback.xml` the following message indicates, that in the c14n-Flow could be my error:

`
2016-02-24 14:56:09,782 - DEBUG [net.shibboleth.idp.authn.impl.SelectSubjectCanonicalizationFlow:100] - Profile Action SelectSubjectCanonicalizationFlow: Checking canonicalization flow c14n/u2f for applicability...

2016-02-24 14:56:10,004 - DEBUG [net.shibboleth.idp.authn.impl.SelectSubjectCanonicalizationFlow:83] - Profile Action SelectSubjectCanonicalizationFlow: Selecting canonicalization flow c14n/u2f

2016-02-24 14:56:10,088 - DEBUG [net.shibboleth.idp.session.impl.DetectIdentitySwitch:148] - Profile Action DetectIdentitySwitch: No previous session found, nothing to do

2016-02-24 14:56:10,106 - DEBUG [net.shibboleth.idp.authn.impl.FinalizeAuthentication:137] - Profile Action FinalizeAuthentication: Canonical principal name was established as 'null'

2016-02-24 14:56:10,112 - DEBUG [net.shibboleth.idp.authn.impl.FinalizeAuthentication:186] - Profile Action FinalizeAuthentication: Request did not have explicit authentication requirements, result is accepted
`

I am not sure if I have to tweak my configuration or if it is a bug.
Thank you and kind regards,
Martin

**Comment by Ratler** [comment]
@gegmar could you add `<logger name="eu.stderr" level="DEBUG" />` in `conf/logback.xml` and try again? Might give us a clue what is going on.

**Comment by gegmar** [comment]
Here is my log-dump.

webidp.test.at = Shibboleth IdP
kerberos.test.at = LDAP-UserDir + U2FVal-Server (http://kerberos.test.at/wsapi/u2fval/)
mail.test.at = Shibboleth SP
User = user2
[log-dump.txt](https://github.com/Ratler/shibboleth-mfa-u2f-auth/files/146167/log-dump.txt)

Regards, Martin

**Comment by Ratler** [comment]
Thanks for the log. It looks like the username is actually correctly carried through the user context. Not sure why it would fail to populate the C14N (since it appear to be loaded configure wise). I will be back home tonight which means I will have some time to really dig into it. I will let you know later today if it's a configuration issue or something else.

**Comment by gegmar** [comment]
Thank you for your effort! Let me know, if you need further information about my setup.

**Comment by gegmar** [comment]
Hi,

I have found a workaround to solve this problem (for me):

I changed in `ValidateU2fResponse.groovy` in Method `populateSubject` the added Principal-class from `U2fPrincipal` to the built-in class `UsernamePrincipal` (also added `import net.shibboleth.idp.authn.principal.UsernamePrincipal` to the imports in head). After this change, c14n/simple will be selected and this flow exports the username.

Maybe this helps find the real issue, why the original way does not work for me.
Regards, Martin
